### PR TITLE
proposal for time-to-full-display span in ui transactions

### PR DESCRIPTION
**Pull request by stefanosiano**
We want to provide a new span to the automatic ui transactions.
The ttfd (time-to-full-display) span is a way for the user to notify the sdk that a ui has been fully loaded.
That is, after all data is retrieved, either by database calls or rest apis, and set into the ui.
We would create a new span to the ui automatic transactions to measure it, for all screens of the application.
This rfc is mostly about how we should design the new api.

Also please edit the "Rendered RFC" link so one can quickly get to the rendered markdown file.

[Rendered RFC](https://github.com/getsentry/rfcs/blob/ui-transaction-ttfd-span/text/0046-ttfd-automatic-transaction-span.md)

### CODE REVIEW

File: text/0046-ttfd-automatic-transaction-span.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+E.g. Activity A starts -> Activity B starts -> Activity A finishes loading data and the api is called.  
+At this point without the activity it was called on, we wouldn't know which span to finish, because the activity B would be at the top of the stack.  
+ - We need to add an api to `SentryAndroid`, instead of the `Sentry` class used everywhere else, due to Activity dependency.  
+ - This is not ideal for single activity apps, as it wouldn't work for fragments.  
+ - If the user doesn't call the api, we would have a span that runs forever. We would have to add a timeout to automatically cancel the span.  
+2) Add a `Sentry.monitorFullyDrawn()` api. We would start the span automatically when an Activity is being created.  
+This api would return the span or a custom object to allow the user to finish it autonomously.  
+  
+Pros:  
+We can flag the span when the api is called, so that if the user doesn't call the api, we know we can cancel it.
```
Comment (philipphofmann) [inline text/0046-ttfd-automatic-transaction-span.md:41]: What do you mean by `we can flag the span`?

File: text/0046-ttfd-automatic-transaction-span.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ - We need the activity this api is called for, and passing an Activity instance to an api is not ideal.  
+E.g. Activity A starts -> Activity B starts -> Activity A finishes loading data and the api is called.  
+At this point without the activity it was called on, we wouldn't know which span to finish, because the activity B would be at the top of the stack.  
+ - We need to add an api to `SentryAndroid`, instead of the `Sentry` class used everywhere else, due to Activity dependency.  
+ - This is not ideal for single activity apps, as it wouldn't work for fragments.  
+ - If the user doesn't call the api, we would have a span that runs forever. We would have to add a timeout to automatically cancel the span.  
+2) Add a `Sentry.monitorFullyDrawn()` api. We would start the span automatically when an Activity is being created.  
+This api would return the span or a custom object to allow the user to finish it autonomously.  
+  
+Pros:
```
Comment (philipphofmann) [inline text/0046-ttfd-automatic-transaction-span.md:40]: `m`: Another pro is that it doesn't depend on Android and can be used on multiple UI platforms.

File: text/0046-ttfd-automatic-transaction-span.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ - We need to add an api to `SentryAndroid`, instead of the `Sentry` class used everywhere else, due to Activity dependency.  
+ - This is not ideal for single activity apps, as it wouldn't work for fragments.  
+ - If the user doesn't call the api, we would have a span that runs forever. We would have to add a timeout to automatically cancel the span.  
+2) Add a `Sentry.monitorFullyDrawn()` api. We would start the span automatically when an Activity is being created.  
+This api would return the span or a custom object to allow the user to finish it autonomously.  
+  
+Pros:  
+We can flag the span when the api is called, so that if the user doesn't call the api, we know we can cancel it.  
+Cons:
+- Returning the span would allow the user to perform "dangerous" operations. We could solve this by returning a stripped interface to allow only the `finish()` method, or an entirely custom object.
```
Comment (philipphofmann) [inline text/0046-ttfd-automatic-transaction-span.md:43]: What are `"dangerous" operations`?

File: text/0046-ttfd-automatic-transaction-span.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ - This is not ideal for single activity apps, as it wouldn't work for fragments.  
+ - If the user doesn't call the api, we would have a span that runs forever. We would have to add a timeout to automatically cancel the span.  
+2) Add a `Sentry.monitorFullyDrawn()` api. We would start the span automatically when an Activity is being created.  
+This api would return the span or a custom object to allow the user to finish it autonomously.  
+  
+Pros:  
+We can flag the span when the api is called, so that if the user doesn't call the api, we know we can cancel it.  
+Cons:
+- Returning the span would allow the user to perform "dangerous" operations. We could solve this by returning a stripped interface to allow only the `finish()` method, or an entirely custom object.  
+- If the user doesn't call the api, we would have a span that runs forever. We would have to add a timeout to automatically cancel the span.
```
Comment (philipphofmann) [inline text/0046-ttfd-automatic-transaction-span.md:44]: `m`: Another con is that we can't reliably map `Sentry.monitorFullyDrawn()` to the correct APM transaction.

File: text/0046-ttfd-automatic-transaction-span.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+2) Add a `Sentry.monitorFullyDrawn()` api. We would start the span automatically when an Activity is being created.  
+This api would return the span or a custom object to allow the user to finish it autonomously.  
+  
+Pros:  
+We can flag the span when the api is called, so that if the user doesn't call the api, we know we can cancel it.  
+Cons:
+- Returning the span would allow the user to perform "dangerous" operations. We could solve this by returning a stripped interface to allow only the `finish()` method, or an entirely custom object.  
+- If the user doesn't call the api, we would have a span that runs forever. We would have to add a timeout to automatically cancel the span.  
+3) Add a `Sentry.monitorFullyDrawn()` and a `Sentry.reportFullyDrawn(ISpan)` api. We would start the span automatically when an Activity is being created.  
+This api would return a uid used by the other api to stop the span.
```
Comment (philipphofmann) [inline text/0046-ttfd-automatic-transaction-span.md:46]: `l`:
```suggestion
This api would return a UUID used by the other api to stop the span.  
```

Comment (philipphofmann) [inline text/0046-ttfd-automatic-transaction-span.md:46]: `m`: `Sentry.reportFullyDrawn(ISpan)` takes an `ISpan` as a parameter, but here we write about a UUID. That's a bit confusing.

**Comment by philipphofmann** [comment]
Futher options

4) Add `monitorFullyDrawn()` and `reportFullyDrawn()` to `ISpan`.
The user gets access to the APM UI transaction by calling `Sentry.span`, calls `span.monitorFullyDrawn` and `span.monitorFullyDrawn()`.

Pros

1.  Correlate fully drawn to correct APM transaction
2. User can add more spans via the same API `Sentry.span`
3. Knowing when to wait for fully drawn

Cons

1. Extra APIs to call
2. Keeping a reference of transaction

5) Add reportFullyDrawn()` to `ISpan`.
The user gets access to the APM UI transaction by calling `Sentry.span`, and calls `span.monitorFullyDrawn()`.

Pros

1. Same as 1-2 of option 4

Cons

1. Same as 1-2 of option 4
2. Opposite of pro of option 4

File: text/0046-ttfd-automatic-transaction-span.md (hunk 1 above)
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md:41]: I mean that we can have a mechanism to store whether the user called the api for the current transaction, so that we know if we can cancel the ttfd span or keep it and let the user finish it when he wants

File: text/0046-ttfd-automatic-transaction-span.md (hunk 5 above)
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md:46]: that should be uuid. My fault, fixed it

File: text/0046-ttfd-automatic-transaction-span.md (hunk 3 above)
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md:43]: Operations like creating a new span as child of the ttfd span, or set a different status.
I'm not sure if that would be a real problem, though, or just paranoia 😅

File: text/0046-ttfd-automatic-transaction-span.md (hunk 4 above)
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md:44]: You're right, even if maybe forcing the user to call it in a specific callback (like onActivityCreated) could work. But that would be another con 😄

File: text/0046-ttfd-automatic-transaction-span.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  
+We have to add a new api to the sdk to allow the user to notify that a ui was fully drawn.  
+  
+# Options Considered
+  
+These options were considered for Android, but the same apply to other sdks, too.  
+1) Add a `SentryAndroid.reportFullyDrawn(Activity)` static method. We would start the span automatically when an Activity is being created and we would finish it when the api is called.  
+  
+Pros:  
+This resembles the system api `Activity.reportFullyDrawn()`, making it immediate to use.
```
[the remark this answers is not in the document]
Comment (armcknight) [reply text/0046-ttfd-automatic-transaction-span.md]: I think what you want to say is "making it *obvious* how to use it"

**Review by armcknight — COMMENTED** [review]
I'm not sure if this is meant to be Android specific, but looking at this through a `sentry-cocoa` lens, I think the API would look like so:

- Add a new option to `SentryOptions` to indicate that TTFD should be measured, and add a new static method `+[Sentry reportFullyDrawn]` to stop the span, with a timeout for if it's never called and some way to call that out to customers in the front end. It could still have child spans attached and otherwise behave as any other span, but the static API means customers don't have to manage the handle to it, which can be complicated to thread through various async APIs they use while initializing their app.
    - This will not include pre-main time. We could add some static API that allows measuring pre-main time and excludes pre-warm pauses. We already have some prewarming logic, but in order to measure the actual pre-main work done by an app, we'd need to instruct users on how to add some code to their main.m for ObjC, or for Swift and SwiftUI apps, how to call the API from a @main-annotated construct or a swift.main file.

**Comment by stefanosiano** [comment]
> * Add a new option to `SentryOptions` to indicate that TTFD should be measured, and add a new static method `+[Sentry reportFullyDrawn]` to stop the span, with a timeout for if it's never called and some way to call that out to customers in the front end

This is basically option 1 of this rfc. But how would you understand the screen the static api was called for?

File: text/0046-ttfd-automatic-transaction-span.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  
+We have to add a new API to the SDK to allow the user to notify that the UI was fully drawn.  
+Also, we need the user to specify the span to finish through a parameter, as it cannot be done automatically.  
+E.g. Activity A starts -> Activity B starts -> Activity A finishes loading data and the API is called.  
+At this point without the activity it was called on, we wouldn't know which span to finish, because the activity B would be at the top of the stack.  
+  
+# Options Considered
+  
+These options were considered for Android, but the same apply to other SDKs, too.  
+1) Add a `SentryAndroid.reportFullyDrawn(Activity)` static method. We would start the span automatically when an Activity is being created and we would finish it when the API is called.
```
Comment (marandaneto) [inline text/0046-ttfd-automatic-transaction-span.md:29]: Added in [API level 19](https://developer.android.com/guide/topics/manifest/uses-sdk-element#ApiLevels)
So this feature wont work < 19

**Comment by marandaneto** [comment]
There's a new API for `reportFullyDrawn` from AndroidX Core IIRC, I don't remember the name nor the link, but might be worth investigating this, allows hooking into the Activity `reportFullyDrawn` via callbacks.

File: text/0046-ttfd-automatic-transaction-span.md [hunk 8]
```diff
@@ -0,0 +1,90 @@
+- Start Date: 2022-12-16
+- RFC Type: feature
+- RFC PR: https://github.com/getsentry/rfcs/pull/46
+- RFC Status: draft
+  
+# Summary.
```
Comment (marandaneto) [inline text/0046-ttfd-automatic-transaction-span.md:6]: Is that going to be a transaction, a span or a measurement?
How is this going to play along with https://docs.sentry.io/platforms/android/performance/instrumentation/automatic-instrumentation/#androids-activity-instrumentation ?
Do we intend to do that for Fragments as well? https://docs.sentry.io/platforms/android/performance/instrumentation/automatic-instrumentation/#androids-fragment-instrumentation

**Comment by marandaneto** [comment]
> There's a new API for `reportFullyDrawn` from AndroidX Core IIRC, I don't remember the name nor the link, but might be worth investigating this, allows hooking into the Activity `reportFullyDrawn` via callbacks.

Here we go https://developer.android.com/reference/androidx/activity/FullyDrawnReporter
Maybe another option would be hooking into this API and calling the `span.finish()` automatically.

File: text/0046-ttfd-automatic-transaction-span.md (hunk 8 above)
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md:6]: The idea is to create a new span to attach to the automatic transactions of activities.
We could also later add the same for fragment transactions as well

File: text/0046-ttfd-automatic-transaction-span.md (hunk 7 above)
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md:29]: Nope, we would just add a new API to our sdk, without calling system functions. I think it's better to let the user have control over it

**Comment by stefanosiano** [comment]
> > There's a new API for `reportFullyDrawn` from AndroidX Core IIRC, I don't remember the name nor the link, but might be worth investigating this, allows hooking into the Activity `reportFullyDrawn` via callbacks.
> 
> Here we go https://developer.android.com/reference/androidx/activity/FullyDrawnReporter Maybe another option would be hooking into this API and calling the `span.finish()` automatically.

Yeah, I had a look at that, even if that library version is still in alpha. 
Thanks for the catch, I'm going to add it to the options with pros and cons, and I'd appreciate some comments on them.
The biggest con i think is that it seems to be intended to be called only the first launch of the first Activity, as it says `You can safely call this method any time after first launch as well, in which case it will simply be ignored.` https://developer.android.com/reference/android/app/Activity.html#reportFullyDrawn()

File: text/0046-ttfd-automatic-transaction-span.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md]: > I think what you want to say is "making it _obvious_ how to use it"

yeah 😄

File: text/0046-ttfd-automatic-transaction-span.md (hunk 8 above)
Comment (marandaneto) [reply text/0046-ttfd-automatic-transaction-span.md:6]: Mobile vitals is based on measurements, don't we intend to leverage our mobile vitals with this?
What if the transaction auto finishes before the user called `Sentry.reportFullyDrawn`?

**Comment by marandaneto** [comment]
> > > There's a new API for `reportFullyDrawn` from AndroidX Core IIRC, I don't remember the name nor the link, but might be worth investigating this, allows hooking into the Activity `reportFullyDrawn` via callbacks.
> > 
> > 
> > Here we go [developer.android.com/reference/androidx/activity/FullyDrawnReporter](https://developer.android.com/reference/androidx/activity/FullyDrawnReporter) Maybe another option would be hooking into this API and calling the `span.finish()` automatically.
> 
> Yeah, I had a look at that, even if that library version is still in alpha. Thanks for the catch, I'm going to add it to the options with pros and cons, and I'd appreciate some comments on them. The biggest con i think is that it seems to be intended to be called only the first launch of the first Activity, as it says `You can safely call this method any time after first launch as well, in which case it will simply be ignored.` [developer.android.com/reference/android/app/Activity.html#reportFullyDrawn()](https://developer.android.com/reference/android/app/Activity.html#reportFullyDrawn())

Found more info here https://android-developers.googleblog.com/2022/10/whats-new-in-jetpack-ads-22.html

> Yeah, I had a look at that, even if that library version is still in alpha

Fair point.

> i think is that it seems to be intended to be called only the first launch of the first Activity

Im not sure about that, it depends what they call app start cold/hot/warm state, maybe they consider app start just by bringing the app from the background which is not even the 1st screen anymore, so it'd be fine to be called or it will be ignored if already, hard to tell without reading the source code/more docs.

File: text/0046-ttfd-automatic-transaction-span.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  
+We have to add a new API to the SDK to allow the user to notify that the UI was fully drawn.  
+Also, we need the user to specify the span to finish through a parameter, as it cannot be done automatically.  
+E.g. Activity A starts -> Activity B starts -> Activity A finishes loading data and the API is called.  
+At this point without the activity it was called on, we wouldn't know which span to finish, because the activity B would be at the top of the stack.  
+  
+# Options Considered
+  
+These options were considered for Android, but the same apply to other SDKs, too.  
+1) Add a `SentryAndroid.reportFullyDrawn(Activity)` static method. We would start the span automatically when an Activity is being created and we would finish it when the API is called.
```
Comment (marandaneto) [inline text/0046-ttfd-automatic-transaction-span.md:29]: `reportFullyDrawn` is an Android API, if our feature is called TTFD, I'd suggest naming it to something similar.
I'd also try to add this API to  `SentryAndroid`, also if possible, to not require `Activity`, so it can't be Android less classes.
Why do we need the activity?

File: text/0046-ttfd-automatic-transaction-span.md (hunk 8 above)
Comment (romtsn) [reply text/0046-ttfd-automatic-transaction-span.md:6]: I feel like we could have both, right? The reason for spans is that we can detect performance issues server-side (which works based on spans). But having it as a measurement wouldn't hurt either?

File: text/0046-ttfd-automatic-transaction-span.md (hunk 9 above)
Comment (stefanosiano) [reply text/0046-ttfd-automatic-transaction-span.md:29]: > Im not sure about that, it depends what they call app start cold/hot/warm state, maybe they consider app start just by bringing the app from the background which is not even the 1st screen anymore, so it'd be fine to be called or it will be ignored if already, hard to tell without reading the source code/more docs.

I tried to call `Activity.reportFullyDrawn()` multiple times in multiple activities, and it gets called once per activity, so it would be fine. There is still the question about single activity apps, tough. However, i updated the rfc.

the name would be `time-to-full-display`. The activity is needed (in some options) because we wouldn't know which activity is the one that was fully drawn, in case two activities are started and the previous one is not destroyed.

File: text/0046-ttfd-automatic-transaction-span.md (hunk 3 above)
Comment (philipphofmann) [reply text/0046-ttfd-automatic-transaction-span.md:47]: I think that's not a problem, but a feature.

File: text/0046-ttfd-automatic-transaction-span.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (philipphofmann) [reply text/0046-ttfd-automatic-transaction-span.md]: Yes please rephrase it to what @armcknight pointed out.

**Review by philipphofmann — COMMENTED** [review]
We decided on naming the spans for TTFD and TTID, `ui.load.initial_display` and `ui.load.full_display`. I think our API should reflect that naming decision. 

> Report to the system that your app is now fully drawn, for diagnostic and optimization purposes. The system may adjust optimizations to prioritize work that happens before reportFullyDrawn is called, to improve app startup. Misrepresenting the startup window by calling reportFullyDrawn too late or too early may decrease application and startup performance. - https://developer.android.com/reference/android/app/Activity.html#reportFullyDrawn()

When reading the docs above on `reportFullyDrawn` I think you can only use it for the first Activity when the app is fully drawn. So I would agree with @stefanosiano.

**Comment by philipphofmann** [comment]
We agreed on adding an API `Sentry.reportFullDisplayed` (name to be defined), which will try to find the last active screen load transaction and set the time for full-displayed. The active screen load transaction needs to wait for a to-be-defined timeout if the user calls this API. We still need to spec out all the edge cases.

We didn't go with any of the other options because we don't think that users will use the more complicated API. `Sentry.reportFullDisplayed` seems simple enough for users to try. If we find out users are not happy with the API cause it has some problems with multiple screens we are happy cause it means they are using the API.

Furthermore, the SDKs should somehow automatically detect fully drawn, cause we see a higher adoption with that on JavaScript. Users are reluctant to use manual APIs.

We can add an option to enable or disable this behavior.

Participants: @kahest, @AbhiPrasad, @stefanosiano, @philipphofmann

**Comment by gggritso** [comment]
Extremely late feedback, but out of curiosity, why a span? Time-to-full-display sounds like it's a good fit for a measurement. The Performance team has gotten feedback before that spans are not appropriate for these use cases, since they don't represent a specific unit of work. Plus, with measurements, users would be able to search and filter transactions in Discover and so on. Long time-to-full-display spans might show up as suspect spans, would show up in Performance span breakdowns, etc.

**Comment by stefanosiano** [comment]
> why a span?

One of the main reasons was to allow the usage of Time-to-full-display for performance issues, which are built over spans.
Of course, adding a measurement other than a span makes sense, and we could add both

**Comment by mjq-sentry** [comment]
> One of the main reasons was to allow the usage of Time-to-full-display for performance issues, which are built over spans.
Of course, adding a measurement other than a span makes sense, and we could add both

Is there anything preventing you from creating performance issues using the measurement alone? You have access to the entire event when creating an issue. We use the FCP measurement as a factor in the Large Render Blocking Asset detector, for example.

**Comment by stefanosiano** [comment]
Just throwing an update: 
we are tracking the work on adding `time-to-initial-display` and `time-to-full-display` as measurements [here](https://github.com/getsentry/team-mobile/issues/86)
